File: Project.Trunk/System.Languages/src/php/System.php
<!DOCTYPE html>
<html lang="en">
<head>

</head>
<body style="font-family: 'Source Code Pro', 'Courier New', Helvetica, Geneva">
 <h1>Matrix Utils tests</h1>

<?php
/**
 * Author: Olivier LeDiouris
 * Date: 10/5/18
 * Time: 9:13 AM
 */

class SquareMatrix {
    private $dimension = 0;
    private $matrixElements = array();

    function SquareMatrix($dim, $init) {
        $this->dimension = $dim;
        if ($init) {
            $this->matrixElements = array();
            for ($row = 0; $row < $this->dimension; $row++) {
                $this->matrixElements[$row] = array();
                for ($col = 0; $col < $this->dimension; $col++) {
                    $this->matrixElements[$row][$col] = 0;
                }
            }
        }
    }

    function getDim() {
        return $this->dimension;
    }

    function getElements() {
        return $this->matrixElements;
    }

    function setElementAt($row, $col, $val) {
        $this->matrixElements[$row][$col] = $val;
    }

    function getElementAt($row, $col) {
        return $this->matrixElements[$row][$col];
    }
}

class MatrixUtil {

    const DEBUG = false;

    public static function printMatrix($m) {
        $elements = $m->getElements();
        $elem = "";
        $elem .= "<table>";
        for ($row=0; $row<count($elements); $row++) {
            $elem .= "<tr>";
            $elem .= "<td>|</td>";
            for ($col=0; $col<count($elements[$row]); $col++) {
                $elem .= "<td>";
                $elem .= $elements[$row][$col];
                $elem .= "</td>";
            }
            $elem .= "<td>|</td>";
            $elem .= "</tr>";
        }
        $elem .= "</table>";
        return $elem;
    }

    public static function minor($m, $row, $col) {
        $small = new SquareMatrix($m->getDim() - 1, true);
        for ($c=0; $c<$m->getDim(); $c++) {
            if ($c != $col) {
                for ($r=0; $r<$m->getDim(); $r++) {
                    if ($r != $row) {
                        $small->setElementAt((($r < $row) ? $r : ($r - 1)), (($c < $col) ? $c : ($c - 1)), $m->getElementAt($r, $c));
                    }
                }
            }
        }
        return $small;
    }

    public static function determinant($m) {
		$v = 0.0;

		if ($m->getDim() == 1) {
			$v = $m->getElementAt(0, 0);
		} else {
			// C : column in Major
			for ($C = 0; $C < $m->getDim(); $C++) { // Walk thru first line
				// Minor's determinant
				$minDet = MatrixUtil::determinant(MatrixUtil::minor($m, 0, $C));
				$v += ($m->getElementAt(0, $C) * $minDet * ((-1.0) ** ($C + 1 + 1))); // line C, column 1
			}
		}
		if (self::DEBUG) {
            echo "Determinant of ", MatrixUtil::printMatrix($m), " is ", $v, "<br/>";
        }

		return $v;
	}

	public static function comatrix($m) {
        $co = new SquareMatrix($m->getDim(), true);
        for ($r = 0; $r < $m->getDim(); $r++) {
            for ($c = 0; $c < $m->getDim(); $c++) {
                $co->setElementAt($r, $c, MatrixUtil::determinant(MatrixUtil::minor($m, $r, $c)) * ((-1) ** ($r + $c + 2)));  // r+c+2 = (r+1) + (c+1)...
            }
		}
		if (self::DEBUG) {
            echo "Comatrix:", MatrixUtil::printMatrix($co);
        }
		return $co;
    }

    public static function transposed($m) {
		$t = new SquareMatrix($m->getDim(), true);
		// Replace line with columns.
		for ($r = 0; $r < $m->getDim(); $r++) {
			for ($c = 0; $c < $m->getDim(); $c++) {
				$t->setElementAt($r, $c, $m->getElementAt($c, $r));
			}
		}
		if (self::DEBUG) {
			echo "Transposed:", MatrixUtil::printMatrix($t);
		}
		return $t;
	}

	public static function multiply($m, $n) {
		$res = new SquareMatrix($m->getDim(), false);
		for ($r = 0; $r < $m->getDim(); $r++) {
			for ($c = 0; $c < $m->getDim(); $c++) {
				$res->setElementAt($r, $c, $m->getElementAt($r, $c) * $n);
			}
		}
		return $res;
	}

	public static function equals($a, $b) {
		if ($a->getDim() != $b->getDim()) {
			return false;
		}
		for ($r=0; $r<$a->getDim(); $r++) {
			for ($c=0; $c<$a->getDim(); $c++) {
				if ($a->getElementAt($r, $c) != $b->getElementAt($r, $c)) {
					return false;
				}
			}
		}
		return true;
	}

	public static function invert($m) {
		return MatrixUtil::multiply(MatrixUtil::transposed(MatrixUtil::comatrix($m)), (1.0 / MatrixUtil::determinant($m)));
	}

	public static function solveSystem($m,
	                                   $c) {
		$result = array();

		$inv = MatrixUtil::invert($m);

		// Print inverted Matrix
		if (self::DEBUG) {
			echo "Inverted:", MatrixUtil::printMatrix($inv);
		}

		// Lines * Column
		for ($row = 0; $row < $m->getDim(); $row++) {
			$result[$row] = 0.0;
			for ($col = 0; $col < $m->getDim(); $col++) {
				$result[$row] += ($inv->getElementAt($row, $col) * $c[$col]);
			}
		}
		return $result;
	}
}

echo "<hr/>";

$matrix = new SquareMatrix(3, true);
$matrix->setElementAt(0, 0, 12);
$matrix->setElementAt(0, 1, 13);
$matrix->setElementAt(0, 2, 14);

$matrix->setElementAt(1, 0, 1.345);
$matrix->setElementAt(1, 1, -654);
$matrix->setElementAt(1, 2, 0.001);

$matrix->setElementAt(2, 0, 23.09);
$matrix->setElementAt(2, 1, 5.3);
$matrix->setElementAt(2, 2, -12.34);

echo "Matrix dim: ", $matrix->getDim();
echo "<br/>";
$elements = $matrix->getElements();
echo "<br/>";
echo "Det: ", MatrixUtil::determinant($matrix);
echo "<br/>";

echo "<hr/>";
echo "Matrix";
echo MatrixUtil::printMatrix($matrix);
echo "<hr/>";
$constants = array(234, 98.87, 9.876);
$coeffs = MatrixUtil::solveSystem($matrix, $constants);

// Results
echo "System Resolution:<br/>";
for ($i=0; $i<count($coeffs); $i++) {
    echo "Coeff, deg ", (count($coeffs) - 1 - $i), " : ", $coeffs[$i], "<br/>";
}

?>

</body>
</html>

File: Project.Trunk/System.Languages/src/main/C/system.c
/*
 * Compile with
 * gcc -o system system.c
 ***********************************************************************
 * Oliv proudly did it.
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

#define TRUE 1
#define FALSE 0

#define DEBUG FALSE

typedef struct ALLOCATED_PTRS {
	struct ALLOCATED_PTRS * next;
	void * ptr;
} AllocatedPtrs;

#define walkList(a,b) for (b=a;b;b=b->next)

AllocatedPtrs * ptrs = NULL;

// Alloc/Free utilities
void chainAppend (AllocatedPtrs * head, AllocatedPtrs * pt) {
  AllocatedPtrs * structPt = head;

  while (structPt->next) {
    structPt = structPt->next;
  }
  structPt->next = pt;
}

void freeChain (AllocatedPtrs * head) {
  AllocatedPtrs * nextOne;
  AllocatedPtrs * thisOne;

  thisOne = head;
  while (thisOne) {
    nextOne = thisOne->next;
    free((char *)thisOne);
    thisOne = nextOne;
  }
}

void chainRemove (AllocatedPtrs * head, AllocatedPtrs * pt) {
  AllocatedPtrs * locPt;

  walkList(head, locPt) {
    if (locPt->next == pt) {
      locPt->next = pt->next;
      break;
    }
  }
}

void addPtrToList(void * ptr) {
	AllocatedPtrs * thisPtr = (AllocatedPtrs *)malloc(sizeof(AllocatedPtrs));
	thisPtr->ptr = ptr;
	thisPtr->next = NULL;
	if (ptrs == NULL) {
		ptrs = thisPtr;
	} else {
		chainAppend(ptrs, thisPtr);
	}
}

// System & Matrix utilities

double * minorMat(int dim, double * matrix, int row, int col) {
	double * small = (double *) malloc((dim - 1) * (dim - 1) * sizeof(double));
	for (int c=0; c<dim; c++) {
		if (c != col) {
			for (int r=0; r<dim; r++) {
				if (r != row) {
					small[(((r < row) ? r : (r - 1)) * (dim - 1)) + ((c < col) ? c : (c - 1))] = matrix[(r * dim) + c];
				}
			}
		}
	}
	addPtrToList(small);
	return small;
}

double determinant(int dim, double * matrix) {
	double v = 0;
	if (dim == 1) {
		v = matrix[0];
	} else {
		// C: column in major
		for (int C=0; C<dim; C++) { // Walk thru first line
			double minDet = determinant(dim - 1, minorMat(dim, matrix, 0, C));
			v += (matrix[C] * minDet * pow(-1, C + 1 + 1));
		}
	}
	return v;
}

double * comatrix(int dim, double * matrix) {
	double * comat = (double *) malloc(dim * dim * sizeof(double));
	for (int r=0; r<dim; r++) {
		for (int c=0; c<dim; c++) {
			comat[(r * dim) + c] = determinant((dim - 1), minorMat(dim, matrix, r, c)) * pow(-1, (r + c + 2));
		}
	}
	addPtrToList(comat);
	return comat;
}

double * transposed(int dim, double * matrix) {
	double * transp = (double *) malloc(dim * dim * sizeof(double));
	for (int r=0; r<dim; r++) {
		for (int c=0; c<dim; c++) {
			transp[(r * dim) + c] = matrix[(c * dim) + r];
		}
	}
	addPtrToList(transp);
	return transp;
}

double * multiply(int dim, double * matrix, double n) {
	double * mult = (double *) malloc(dim * dim * sizeof(double));
	for (int r=0; r<dim; r++) {
		for (int c=0; c<dim; c++) {
			mult[(r * dim) + c] = matrix[(r * dim) + c] * n;
		}
	}
	addPtrToList(mult);
	return mult;
}

double * invert(int dim, double * matrix) {
	return multiply(dim, transposed(dim, comatrix(dim, matrix)), (1.0 / determinant(dim, matrix)));
}

void printMatrix(int dim, double * matrix) {
	for (int r=0; r<dim; r++) {
		fprintf(stdout, "| ");
		for (int c=0; c<dim; c++) {
			fprintf(stdout, "%lf ", matrix[(r * dim) + c]);
		}
		fprintf(stdout, " |\n");
	}
}

double * solveSystem(int dim, double * matrix, double * coeff) {
	double * result = (double *) malloc(dim * sizeof(double));
	double * inverted = invert(dim, matrix);

	addPtrToList(result);

	if (DEBUG) {
		fprintf(stdout, "Inverted:\n");
		printMatrix(3, inverted);
	}

	for (int r=0; r<dim; r++) {
		result[r] = 0.0;
		for (int c=0; c<dim; c++) {
			result[r] += (inverted[(r * dim) + c] * coeff[c]);
		}
	}
	return result;
}

void printSystem(int dim, double * matrix, double * coeff) {
	char * unknowns = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
	for (int r=0; r<dim; r++) {
		for (int c=0; c<dim; c++) {
			fprintf(stdout, "%s(%lf x %c)", (c==0?"":" + "), matrix[(r * dim) +c], unknowns[c]);
		}
		fprintf(stdout, " = %lf\n", coeff[r]);
	}
}

int main(int argc, char **argv) {

	double matrix[9] = {
		 12,       13,      14,
		  1.345, -654,       0.001,
		 23.09,     5.3, -12.34
	};
	double constants[3] = { 234, 98.87, 9.876 };

	fprintf(stdout, "Solving system:\n");
	printSystem(3, &matrix[0], &constants[0]);
	double * result = solveSystem(3, &matrix[0], &constants[0]);
	fprintf(stdout, "A: %lf\n", result[0]);
	fprintf(stdout, "B: %lf\n", result[1]);
	fprintf(stdout, "C: %lf\n", result[2]);

	// Cleanup
	if (ptrs != NULL) {
		AllocatedPtrs * locPt;
		int nb = 0;
  		walkList(ptrs, locPt) {
			nb++;
			free((char *)locPt->ptr);
		}
		freeChain(ptrs);
		fprintf(stdout, "-- Freed %d pointers\n", nb);
	}
}
